### PR TITLE
Lattice discretization with degenerate and recursive local neighbor symmetry algorithms

### PR DESCRIPTION
**Pull request by GiudGiud**
This PR introduces automatic lattice discretization using either the degenerate and a recursive implementation of local neighbor symmetry. Local neighbor symmetry and degenerate are algorithms laid down in @wbinventor's Phd and the subsequent [article](https://doi.org/10.1016/j.anucene.2018.11.017) to discretize geometries for the purpose of multi group cross section generation. Essentially, degenerate is per-pin discretization, and LNS is per-pin neighborhood discretization. A neighborhood can be "a guide tube on the diagonal, fuel pins on every other side" or "side adjacent to a guide tube and diagonal to another"

The recursive implementation is meant to address one of the shortcomings of local neighbor symmetry, which was not taking into account lattice neighbors when discretizing an assembly within a full core lattice. This was a significant issue for assemblies near the PWR baffle, where fuel pins on one side see other fuel pins, and on the other see the baffle and the reflector. 

Note that because this new routine is using the clone feature, the number of cells involved blows up. Introducing the "clone_regions" boolean in the clone routines limits this burden. Introducing the "clone_materials" boolean in the clone routines limits this burden, fixes issue #899 and allows the creation of reasonably sized MGXS per material libraries. 

This PR has been tested on BEAVRS benchmark individual assemblies and symmetrized (option is_symmetric at input generation) rods-out full core. It yielded the discretizations shown below. The BEAVRS full core also required renaming the baffle universes and most of the assemblies. The reason for that is that even though they are the same design of assembly, they are named differently in the model. 

The "name " attribute option is to allow a little more flexibility with geometries that are symmetric, but are modeled in a way that doesn't look symmetric when just looking at the universe ids. One example is the baffle universes in BEAVRS, which are defined using rotations of one quarter. The ids are all different because in the main lattice the other 3 corners are modeled with extra universes containing cells with the rotation containing the baffle universes of the other corner. Renaming universes and using LNS with the name attributes allows to get past that. 

3.1% enriched 20 burnable poisons BEAVRS Assembly:
xy_plot
Degenerate discretization of cells (above)
xy_plot_mat
Degenerate discretization of materials (only fuel selected) (above)
xy_plot_lns
LNS discretization of cells (above)
xy_plot_mat_lns
LNS discretization of materials (only fuel selected) (above)

Full core symmetric BEAVRS
xy_plot_17
LNS discretization of cells (above). Note that symmetry and rotations incur an extra universe to hold the transformation and therefore dont show up with the same color in plots
xy_plot_mat17
LNS discretization of materials (above). LNS allows the full core to be represented with only 390 fuel materials, instead of >50,000 for a degenerate discretization

### CODE REVIEW

**Comment by GiudGiud** [comment]
Thanks for the review !

The problem with that case is that the neighbors of each sublattice are rotated, yet the lattice is 
-not invariant by rotations (unlike most assemblies)
-not rotated in a way such that the geometry was even symmetric in the first place
This is not the typical use case for LNS, where the system exhibits symmetry, and we want to capture it with the discretization. (see the assemblies with BP loadings on the periphery of BEAVRS, which have rotated neighbor patterns, and are rotated themselves)

I'll see if I can find a fix for those false positives. 
The lack of inner lattice discretization of your example is also concerning. I'll look into it

File: openmc/lattice.py [hunk 1]
```diff
@@ -753,6 +764,313 @@ def is_valid_index(self, idx):
...
+            # Create a clone of the universe, without cloning materials
+            new_universe = self.universes[first_pos[1]][first_pos[0]].clone(
+                 clone_materials=False, clone_regions=False)
+
+            # Call LNS on the immediate sub lattices of this universe
+            for cell_id, sub_cell in new_universe.cells.items():
+
+                sub_universe = sub_cell.fill
+
+                try:
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply openmc/lattice.py]: Went for attributeError in case discretize routines are added to other classes

File: openmc/lattice.py [hunk 2]
```diff
@@ -753,6 +764,336 @@ def is_valid_index(self, idx):
...
+                    else:
+                        pattern[2, :] = dummy_neighbor
+                else:
+                    if i > 0:
+                        pattern[2, 0] = getattr(self.universes[j+1][i-1],
+                                                attribute)
+                    pattern[2, 1] = getattr(self.universes[j+1][i], attribute)
+                    if i < self.shape[0] - 1:
+                        pattern[2, 2] = getattr(self.universes[j+1][i+1],
+                                                attribute)
```
Comment (smharper) [inline openmc/lattice.py]: All of these nested `if`s to handle the special cases make the code kind of tough to read. And it will get worse if anyone extends this to 3D. I think something you can do to simplify the code would be to write two locally defined functions (define them inside this function itself) that take `i` and `j` values and return the appropriate universe. One function would handle the case for `len(lattice_neighbors) > 0` and another would handle the other case. Then you can just check `len(lattice-neighbors)` once and alias the appropriate function for use in the `for` loops.

File: openmc/lattice.py [hunk 3]
```diff
@@ -753,6 +764,336 @@ def is_valid_index(self, idx):
...
+        for pattern, pattern_data in patterns.items():
+
+            first_pos = pattern_data['locations'][0]
+
+            # Create a clone of the universe, without cloning materials
+            new_universe = self.universes[first_pos[1]][first_pos[0]].clone(
+                 clone_materials=False, clone_regions=False)
+
+            # Call LNS on the immediate sub lattices of this universe
+            for cell_id, sub_cell in new_universe.cells.items():
```
Comment (smharper) [inline openmc/lattice.py]:
```suggestion
            for sub_cell in new_universe.cells.values():
```

File: openmc/universe.py [hunk 4]
```diff
@@ -477,12 +477,18 @@ def get_all_universes(self):
...
         return universes
 
-    def clone(self, memo=None):
+    def clone(self, clone_materials=True, clone_regions=True, memo=None):
         """Create a copy of this universe with a new unique ID, and clones
         all cells within this universe.
 
         Parameters
         ----------
+        clone_materials : boolean
```
Comment (smharper) [inline openmc/universe.py]:
```suggestion
        clone_materials : bool
```

File: openmc/universe.py [hunk 5]
```diff
@@ -477,12 +477,18 @@ def get_all_universes(self):
...
+            Whether to create separates copies of the materials filling cells
+            contained in this universe. Default is True.
+        clone_regions : boolean
```
Comment (smharper) [inline openmc/universe.py]:
```suggestion
        clone_regions : bool
```

File: openmc/universe.py [hunk 6]
```diff
@@ -477,12 +477,18 @@ def get_all_universes(self):
...
+            Whether to create separates copies of the regions bounding cells
+            contained in this universe. Default is True.
```
Comment (smharper) [inline openmc/universe.py]: Remove `Default is True`

File: tests/unit_tests/test_lattice_discretization.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+import xml.etree.ElementTree as ET
+
+import openmc
+import pytest
+
+from tests.unit_tests.test_lattice import zr, pincell1, pincell2, rlat2
+
+
+def test_discretization_degenerate(rlat2):
+
```
Comment (smharper) [inline tests/unit_tests/test_lattice_discretization.py]: Remove the unnecessary whitespace after the function definition here and elsewhere in the tests functions you've added

File: openmc/lattice.py (hunk 2 above)
Comment (GiudGiud) [reply openmc/lattice.py]: Agreed! It's much nicer that way. 
I didnt exactly create a separate function of the case that neighbors arent specified, I made:
 - one function for neighbors outside the lattice, which got the len(lattice_neighbors) if statement out
- one function for neighbors inside the lattice

File: openmc/lattice.py [hunk 8]
```diff
@@ -753,6 +764,336 @@ def is_valid_index(self, idx):
...
+        attribute : String
+            Which universe attribute to use when detecting neighbor symmetry
+            patterns. Can be the 'id' or the 'name' of the lattice universes
+
+        """
+
+        # Check routine inputs
+        if self.ndim == 3:
+            raise NotImplementedError("LNS discretization is not implemented "
+                                      "for 3D lattices")
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply openmc/lattice.py]: done. It probably would work with minor changes but I dont think there's a need

File: openmc/universe.py (hunk 6 above)
Comment (GiudGiud) [reply openmc/universe.py]: done

File: tests/unit_tests/test_lattice_discretization.py (hunk 7 above)
Comment (GiudGiud) [reply tests/unit_tests/test_lattice_discretization.py]: done

File: openmc/lattice.py [hunk 9]
```diff
@@ -753,6 +764,336 @@ def is_valid_index(self, idx):
...
+                    continue
+
+                # Form the pattern at and around each universe
+                if attribute == "id":
+                    pattern = np.empty(shape=(3, 3), dtype=np.int)
+                if attribute == "name":
+                    pattern = np.empty(shape=(3, 3), dtype=np.dtype('U100'))
+                    if len(self.universes[j][i].name) > 100:
+                        warning.warn("Universe name is larger than 100 "
+                             "characters, which may hinder neighbor search.")
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply openmc/lattice.py]: That would solve the issue but then I probably would need to hash the result of getattr as well. Even when using the id attribute. It'll make the code harder to debug if we ever need to

**Comment by smharper** [comment]
The `discretize` function is not super general. For example, it will only discretize the lattice of pins if that lattice is just one coordinate level below the assembly lattice. You might have more than one coordinate level in order to add rotations (e.g. making a BWR geometry with control blades). And I think the usefulness is limited by how difficult it is to figure out where in the geometry each cell lies. A user would have to do a lot of forensics to figure out which part of the geometry corresponds to their MGXS outputs.

What do you think about just adding this feature to the BEAVRS repo as opposed to OpenMC itself? (With an OpenMC PR for the cloning stuff.)

**Comment by GiudGiud** [comment]
This first point could be addressed in another commit.
The limitation to one-level-below was to avoid worrying about discretizing the same lattice twice (for example for a lattice two levels down, once in the first call, once in a recursive call). It was also meant to avoid the additional complexity that layers of rotations could introduce. The list of neigbors would need to be 'rotated', really nothing that can't be done in this framework. However, it's just wiser to wait for someone to actually have the need for the feature.

Digging in the rotations of universes[x ,y] in a lattice when looking for neighbors could actually remove the need for the "rotate_universe_with_neighbors" boolean. Simply put, the universe would never be rotated, unless it contains a cell that is rotated, and the universe inside that cell matches another universe in the lattice.
This is a reasonable assumption about how the geometry was constructed. 

With regards to finding where the material actually lies, the materials are still fills of cells. Any per-fuel pin discretization will require you to look for them quite the same way. Any code, such as OpenMOC, that loads the MGXS libraries from OpenMC will know where the material lies, since it knows where the cells are. 
Now, if a code can only load the hdf5 mgxs files, which only contains the material ids and their cross sections, then they already have to input **manually** which materials id corresponds to which construct in their code.
Anyway, I don't think encoding location information in the material is the way forward. The geometry CSG tree is here for that purpose. 

The other option to store cross sections, distributed materials, requires a lot more hard-coding of the assembly shapes and would be even less general. Not to mention exactly the same difficulties you mentioned if using a non obvious discretization.
A PR to BEAVRS would effectively bury the code. The next user who needs some sort of cross section discretization of a reactor with rectangular lattices would have to start from scratch. Cloning fuel pins one by one in a complex CSG tree, making sure that only the fuel material is cloned etc.., is very menial.

**Comment by smharper** [comment]
We talked in-person about some changes to make with this PR. In particular, checking for lattices nested in cells with user-specified rotations.

**Comment by GiudGiud** [comment]
I have a working and tested implementation that checks for user-specified rotations and avoids the issue with BWR control blades we mentioned. 

However, this doesnt work for BEAVRS anymore:
- the core lattice's universes contain two cells: one with the sublattice (assembly), and one with the water gap. We wanted to avoid doing LNS in this situation (two cells in universes[i,j]) since there is no way to check for the symmetry of a cell, only for a lattice it's easy to do. 
- the assembly rotations are not user-defined in BEAVRS, instead there's a completely different lattice defined for every orientation. I was getting around that by renaming it and getting the right rotation from the neighbors around each lattice. 

Leaving this open for now in case anyone wants to pitch in. 
I think the current version is a better compromise with regards to detecting symmetry / avoiding false positives / working on the actual case it's necessary for.

**Comment by GiudGiud** [comment]
Latest commit removes:
- automatic handling of neighbors for sublattices
- automatic rotation of sublattices with their neighbor patterns

While these functionalities were nice, they couldn't handle 
- the case where symmetry was broken because an extra asymmetric cell existed next to the one containing sublattice
- the case where the neighbor pattern was rotated, but the lattice was not symmetric so any rotation (which would be done to match the neighbor pattern) would actually modify the geometry, thus would look like a bug

Because the capability for handling lattice's neighbors remains, this can do a little more than regular LNS and can be used to model BEAVRS, discriminating assembly sides facing the baffle and facing inside.

File: openmc/lattice.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (smharper) [reply openmc/lattice.py]: Okay, then maybe just use `dtype='O'`. That will work for a string of any length, and you won't need to have these different `if` cases for `id` versus `name`.

Also, you can move this initialization of `pattern` outside of the `for` loops so that it's only done once.

File: openmc/lattice.py [hunk 10]
```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
...
+
+                # Create a neighborhood pattern based on the universe's
+                # neighbors in the grid, and lattice's neighbors at the edges
+
+                # Find neighbors among lattice's neighbors at the edges
+                if len(lattice_neighbors) > 0:
+                    find_edge_neighbors(i, j)
+                else:
+                    pattern[:, :] = dummy_neighbor
```
Comment (smharper) [inline openmc/lattice.py]: You can move this check for `if len(lattice_neighbors) > 0` into `find_edge_neighbors`. That will make all of the lattice edge logic localized into the `find_edge_neighbors` function.

File: openmc/lattice.py [hunk 11]
```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
...
+
+                sub_universe = sub_cell.fill
+
+                try:
+                    sub_universe.discretize(strategy,
+                                            universes_to_ignore,
+                                            materials_to_clone,
+                                            lattice_neighbors=[],
+                                            attribute=attribute)
+                except AttributeError:
```
Comment (smharper) [inline openmc/lattice.py]: I know I suggested `AttributeError`, but a whole lot of things can throw this error so its better to directly check `if hasattr(sub_universe, discretize)` instead of using a `try`

File: openmc/lattice.py [hunk 12]
```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
...
+                    continue
+
+            # Replace only the materials in materials_to_clone
+            for material in materials_to_clone:
+                material_cloned = False
+
+                for cell in new_universe.get_all_cells().values():
+
+                    if cell.fill is not None:
+                        if cell.fill.id == material.id:
```
Comment (smharper) [inline openmc/lattice.py]: The cell fill could be a `Universe` so you probably need to check that this fill is a `Material`.

Also, there will be an exception here for cells with distribmats. I'm not really sure how LNS should handle distribmats so you should probably just raise a `ValueError` if a distribmat is detected in the lattice.

File: openmc/lattice.py [hunk 13]
```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
...
+        materials_to_clone : Iterable of Material
+            List of materials that should be cloned when discretizing
+        lattice_neighbors : Iterable of Integral or str
+            List of attributes of the lattice's neighbors. By default, if
+            present, the lattice outer universe will be used. The neighbors
+            are represented as follows [top left, top, top right, left, right,
+            bottom left, bottom, bottom right]
+        attribute : str
+            Which universe attribute to use when detecting neighbor symmetry
+            patterns. Can be the 'id' or the 'name' of the lattice universes
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply openmc/lattice.py]: This is brilliant, switching to using a key function will allow enough flexibility to discretize models where symmetries are only known to the user.
Working on it

File: openmc/lattice.py [hunk 14]
```diff
@@ -759,6 +770,277 @@ def is_valid_index(self, idx):
...
+            elif attribute == 'name':
+                dummy_neighbor = "dummy"
+
+        # Dictionary that will keep track of where each pattern appears, how
+        # it was rotated and/or symmetrized
+        patterns = {}
+
+        # Define an auxiliary function that returns a universe's neighbors
+        # that are outside the lattice
+        def find_edge_neighbors(i, j):
```
[the remark this answers is not in the document]
Comment (GiudGiud) [reply openmc/lattice.py]: ok will do

File: openmc/lattice.py (hunk 11 above)
Comment (GiudGiud) [reply openmc/lattice.py]: I'm getting rid of the recursive call. I think in a lot of cases, the 'key' function will only be valid at a certain lattice level, therefore passing it down with recursion wont work

File: openmc/lattice.py (hunk 10 above)
Comment (GiudGiud) [reply openmc/lattice.py]: done. You're right it makes more sense

File: openmc/lattice.py (hunk 12 above)
Comment (GiudGiud) [reply openmc/lattice.py]: Added. Also added an error for distribcells, detected through the length of their temperature and fill attributes.

**Comment by smharper** [comment]
Looks good!
